pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2366,6 +2404,23 @@ public void endReplaceSegments(String tableNameWithType, String segmentLineageEn
         tableNameWithType, segmentLineageEntryId);
   }
 
+  private Set<String> getOnlineSegmentsFromExternalView(String tableNameWithType) {
+    ExternalView externalView = getTableExternalView(tableNameWithType);
+    if (externalView != null) {
+      Set<String> partitionSet = externalView.getPartitionSet();
+      Set<String> onlineSegments = new HashSet<>(HashUtil.getHashMapCapacity(partitionSet.size()));
+      for (String partition : partitionSet) {
+        Map<String, String> instanceStateMap = externalView.getStateMap(partition);
+        if (instanceStateMap.containsValue(SegmentStateModel.ONLINE) || instanceStateMap
+            .containsValue(SegmentStateModel.CONSUMING)) {
+          onlineSegments.add(partition);
+        }
+      }
+      return onlineSegments;
```

Comment:
Use `externalView.getMapFields()` to avoid extra map lookups.
```suggestion
      Map<String, Map<String, String>> segmentAssignment = externalView.getMapFields();
      Set<String> onlineSegments = new HashSet<>(HashUtil.getHashMapCapacity(segmentAssignment.size()));
      for (Map.Entry<String, Map<String, String>> entry : segmentAssignment.entrySet()) {
        Map<String, String> instanceStateMap = entry.getValue();
        if (instanceStateMap.containsValue(SegmentStateModel.ONLINE) || instanceStateMap
            .containsValue(SegmentStateModel.CONSUMING)) {
          onlineSegments.add(entry.getKey());
        }
      }
      return onlineSegments;
```

Reply:
@Jackie-Jiang `ExternalView` does not expose an API to directly fetch the hash map.

Reply:
fixed